File: quick_sort/quick_sort.go
package main

import (
	"fmt"
)

var arr = []int{20, 43, 52, -1, 43, 29, 34}

func main() {
	fmt.Println("Unsorted: ", arr)
	quick_sort(arr)
	fmt.Println("Sorted: ", quick_sort(arr))
}

func quick_sort(arr []int) []int {
	var recurse func(left int, right int)
	var swap func(i int, j int)
	var partition func(left int, right int, pivot int) int

	swap = func(i int, j int) {
		var temp = arr[i]
		arr[i] = arr[j]
		arr[j] = temp
	}

	partition = func(left int, right int, pivot int) int {
		v := arr[pivot]
		right--
		swap(pivot, right)

		for i := left; i < right; i++ {
			if arr[i] <= v {
				swap(i, left)
				left++
			}
		}

		swap(left, right)
		return left
	}

	recurse = func(left int, right int) {
		if left < right {
			pivot := (right + left) / 2
			pivot = partition(left, right, pivot)
			recurse(left, pivot)
			recurse(pivot+1, right)
		}
	}

	recurse(0, len(arr))
	return arr
}

File: quick_sort/quick_sort.js
function quick_sort(arr) {

  function partition(left, right, pivot) {
    var v = arr[pivot];
    swap(pivot, --right);

    for (var i = left;  i < right; i ++) {
      console.log(arr, left, right, i, arr[i], v);
      if (arr[i] <= v) {
        swap(i, left++);
      }
    }
    swap(left, right);

    return left;
  }

  function swap(i, j) {
    var temp = arr[i];

    arr[i] = arr[j];
    arr[j] = temp;
  }

  function recurse(left, right) {
    if (left < right) {
      var pivot = ~~((left + right) / 2)
      pivot = partition(left, right, pivot);
      recurse(left, pivot);
      recurse(pivot + 1, right);
    }
  }

  recurse(0, arr.length)
  return arr;
}

var arr = [20, 43, 52, -1, 43, 29, 34];

console.log(quick_sort(arr));
